File: 백준/Silver/1074. Z/Z.py
N, r, c = map(int, input().split())

answer = 0


def recursion(length, x, y):
    global answer

    # 찾는 행과 열에 도달했을 때
    if x == r and y == c:
        print(int(answer))
        return
    
    # 더 쪼갤 수 없을만큼 쪼갰을때 하나씩 탐색하기(1씩 더하기.)
    if length == 1:
        answer += 1
        return
    
    # r과 c중 하나라도 범위 밖에 있다면 넓이더해주기.
    if not ((x <= r < x+length) and (y <= c < y+length)):
        answer += length**2
        return
    
    # 1 사분면
    recursion(length/2, x, y)
    # 2 사분면
    recursion(length/2, x, y+length/2)
    # 3 사분면
    recursion(length/2, x+length/2, y)
    # 4 사분면
    recursion(length/2, x+length/2, y+length/2)


recursion(2**N, 0, 0)

File: 백준/Silver/1074. Z/Z.js
const readline = require("readline");
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let inputLines = [];
rl.on("line", (line) => {
  inputLines = line.split(" ").map((e) => Number(e));
});

rl.on("close", () => {
  const N = inputLines[0];
  const r = inputLines[1];
  const c = inputLines[2];

  let ans = 0;

  const recursion = (size, x, y) => {
    if (x === r && y === c) {
      console.log(ans);
      return;
    }

    if (size === 1) {
      ans += 1;
      return;
    }

    if (!(x <= r && r < x + size) && !(y <= c && c < y + size)) {
      ans += size * size;
      return;
    }
    recursion(size / 2, x, y);
    recursion(size / 2, x, y + size / 2);
    recursion(size / 2, x + size / 2, y);
    recursion(size / 2, x + size / 2, y + size / 2);
  };

  recursion(2 ** N, 0, 0);
});
